Create ynh_add_config

yalh76 opened the pull request:
Bored of list of variables to replace on apps config file

Used on:
- https://github.com/YunoHost-Apps/ghost_ynh

yalh76 committed:
Create ynh_add_config

yalh76 committed:
fix default origin/destination

yalh76 committed:
Adding __PATH__

yalh76 committed:
typo

yalh76 committed:
Adding DB stuff

yalh76 committed:
Adding checksum information

yalh76 committed:
Splitting Vars replacement

yalh76 committed:
typo

yalh76 commented:
suggestions:
- [x] see to generalise remplacement all `__TOTO__` by `$toto`
- [x] use sed but check what do ynh_replace_string
- [x] send an error if `$toto` doesn't exist when there a `__TOTO__` in the file
- [x] remove default .env
- [x] change origin to Template
- [x] go to /conf by default to find the Template
- [ ] in the future use ynh_replace_var into other helper instead of ynh_replace_string

yalh76 committed:
Merge branch 'master' into ynh_add_config

yalh76 committed:
switch from origin to template

yalh76 committed:
remove default .env

yalh76 committed:
fix

yalh76 committed:
generalize vars remplacement

yalh76 committed:
sed instead of ynh_replace_string

yalh76 committed:
go to /conf by default to find the Template

yalh76 commented:
@YunoHost-Apps/apps-group ynh_add_config and ynh_replace_vars new helper are ready to be reviewed

alexAubin commented on ynh_add_config/ynh_add_config:
```diff
@@ -0,0 +1,135 @@
+    local delimit=@
+    for one_var in "${uniques_vars[@]}"
+    do
+        if test -n "${!one_var:-}"; then
+            match_string="__${one_var^^}__"
+            match_string=${match_string//${delimit}/"\\${delimit}"}
+            replace_string="${!one_var}"
+            replace_string=${replace_string//${delimit}/"\\${delimit}"}
+            sed --in-place "s${delimit}${match_string}${delimit}${replace_string}${delimit}g" "$file"
+        else
+            ynh_die --message="\$$one_var wasn't initialized when trying to replace __${one_var^^}__ in $file"
+        fi
```
```suggestion
        # Validate that one_var is indeed defined
        test -n "${!one_var:-}" || ynh_die --message="\$$one_var wasn't initialized when trying to replace __${one_var^^}__ in $file"
        
        # Escape delimiter in match/replace string
        match_string="__${one_var^^}__"
        match_string=${match_string//${delimit}/"\\${delimit}"}
        replace_string="${!one_var}"
        replace_string=${replace_string//${delimit}/"\\${delimit}"}
        
        # Actually replace (sed is used instead of ynh_replace_string to avoid triggering an epic amount of debug logs)
        sed --in-place "s${delimit}${match_string}${delimit}${replace_string}${delimit}g" "$file"
```

alexAubin commented on ynh_add_config/ynh_add_config:
```diff
@@ -0,0 +1,135 @@
+    # Replace othes variables
+
+    # List other unique (__ __) variables in $file
+    local uniques_vars=( $(grep -o '__[^.]*__' $file | sort --unique | sed "s@__\([^.]*\)__@\L\1@g" ))
```
```suggestion
    local uniques_vars=( $(grep -o '__[A-Z0-9]+__' $file | sort --unique | sed "s@__\([^.]*\)__@\L\1@g" ))
```

I think I would grep for uppercase stuff here, to be consistent with the fact that later we replace `__${one_var^^}_` (i.e. uppercase version of `$one_var`. Otherwise we could have some people naively trying to have `__foo__` (in lowercase) and wondering why the helper doesn't replace it even though this grep catches it ... (well, one could argue that this new code also silently ignore it so idk)

yalh76 replied:
Thanks, I didn't though of that case :)

yalh76 replied:
After some test I had to switch the grep expression  to `__[A-Z0-9_]*__`
- adding missing `_` used for example on `__DB_USER__`
- and `+` give an empty list when `*` give the good list of variables

alexAubin commented on ynh_add_config/ynh_add_config:
```diff
@@ -0,0 +1,135 @@
+# | arg: -t, --template=     - Template config file to use
+# | arg: -d, --destination=    - Destination of the config file
+#
+# The template can be by default the name of a file in the conf directory
+# of a YunoHost Package, a relative path or an absolute path
+# The helper will use the template $template to generate a config file
+# $destination by replacing the following keywords with global variables
+# that should be defined before calling this helper :
+#   __PATH__      by  $path_url
+#   __NAME__    by $app
+#   __NAMETOCHANGE__    by $app
+#   __USER__    by $app
+#   __FINALPATH__    by $final_path
+#   __PHPVERSION__    by $YNH_PHP_VERSION
+#
+# And any dynamic variables that should be defined before calling this helper like:
+#   __DOMAIN__    by $domain
+#   __APP__    by $app
+#   __VAR_1__    by $var_1
+#   __VAR_2__    by $var_2
```
```suggestion
# The template can be by default the name of a file in the conf directory
# of a YunoHost Package, a relative path or an absolute path
# The helper will use the template $template to generate a config file
# $destination by replacing the following keywords with global variables
# that should be defined before calling this helper :
#   __PATH__             by  $path_url
#   __NAME__            by $app
#   __NAMETOCHANGE__    by $app
#   __USER__            by $app
#   __FINALPATH__       by $final_path
#   __PHPVERSION__      by $YNH_PHP_VERSION
#
# And any dynamic variables that should be defined before calling this helper like:
#   __DOMAIN__   by $domain
#   __APP__      by $app
#   __VAR_1__    by $var_1
#   __VAR_2__    by $var_2
```

alexAubin commented on ynh_add_config/ynh_add_config:
```diff
@@ -0,0 +1,135 @@
+#
+# The helper will replace the following keywords with global variables
+# that should be defined before calling this helper :
+#   __PATH__      by  $path_url
+#   __NAME__    by $app
+#   __NAMETOCHANGE__    by $app
+#   __USER__    by $app
+#   __FINALPATH__    by $final_path
+#   __PHPVERSION__    by $YNH_PHP_VERSION
+#
+# And any dynamic variables that should be defined before calling this helper like:
+#   __DOMAIN__    by $domain
+#   __APP__    by $app
+#   __VAR_1__    by $var_1
+#   __VAR_2__    by $var_2
```
```suggestion
#   __PATH__             by  $path_url
#   __NAME__            by $app
#   __NAMETOCHANGE__    by $app
#   __USER__            by $app
#   __FINALPATH__       by $final_path
#   __PHPVERSION__      by $YNH_PHP_VERSION
#
# And any dynamic variables that should be defined before calling this helper like:
#   __DOMAIN__   by $domain
#   __APP__      by $app
#   __VAR_1__    by $var_1
#   __VAR_2__    by $var_2
```

yalh76 committed:
Update ynh_add_config/ynh_add_config

Co-authored-by: Alexandre Aubin <alex.aubin@mailoo.org>

yalh76 committed:
Update ynh_add_config/ynh_add_config

Co-authored-by: Alexandre Aubin <alex.aubin@mailoo.org>

yalh76 committed:
Update ynh_add_config/ynh_add_config

Co-authored-by: Alexandre Aubin <alex.aubin@mailoo.org>

yalh76 committed:
Update ynh_add_config/ynh_add_config

Co-authored-by: Alexandre Aubin <alex.aubin@mailoo.org>

yalh76 committed:
Update ynh_add_config

- `_` was missing in the character selection
- using `+` give an empty list / with `*` the list goes well

alexAubin reviewed (approved):
:+1:
